charlotte_core/src/memory/pmm.rs

```diff
@@ -0,0 +1,243 @@
+        self.entries.iter()
+    }
+    pub fn find_best_fit(&self, size: usize) -> Result<&bootinfo::memory_map::Entry, Error> {
+        self.entries
+            .iter()
+            .filter(|entry| entry.entry_type == bootinfo::memory_map::EntryType::USABLE)
+            .filter(|entry| entry.length as usize >= size)
+            .min_by_key(|entry| entry.length)
+            .ok_or(Error::InsufficientContiguousMemoryAvailable)
+            .copied()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Error {
+    OutOfMemory,
+    InsufficientMemoryAvailable,
+    InsufficientContiguousMemoryAvailable,
+    MemoryOvercommitted,
+    AddressMisaligned,
+    AddressOutOfRange,
+    InvalidSize,
+    InvalidAlignment,
+}
+
+enum RegionAvailability {
+    Available,
+    Unavailable(PhysicalAddress),
+}
+
+const FRAME_SIZE: usize = 4096;
+
+/// A bitmap based physical frame allocator
+pub struct PhysicalFrameAllocator {
+    bitmap: &'static mut [u8]
+}
+
+impl PhysicalFrameAllocator {
+    fn new() -> PhysicalFrameAllocator {
+        let memory_map = MemoryMap::get();
+        let total_memory = &memory_map.total_memory();
+        // find a region that is large enough to hold the bitmap
+        let region = &memory_map.find_best_fit(total_memory / FRAME_SIZE / 8)
+            .expect("Failed to find a physical memory region large enough to hold the physical frame allocator bitmap");
+        // create the PFA
+        let pfa = PhysicalFrameAllocator {
+            // Safety: The region is guaranteed to be valid and the length is guaranteed to be large enough to hold the bitmap
+            bitmap: unsafe { slice_from_raw_parts_mut((*DIRECT_MAP.lock().deref_mut() + region.base as usize) as *mut u8, total_memory / FRAME_SIZE as usize).as_mut().unwrap_unchecked()}
+        };
+        // clear the bitmap to mark all frames as unavailable
+        for byte in pfa.bitmap.iter_mut() {
+            *byte = 1;
+        }
+        // clear the bits corresponding to available frames
+        for entry in MemoryMap::get().iter() {
+            if entry.entry_type == bootinfo::memory_map::EntryType::USABLE {
+                let start_frame = entry.base as usize / FRAME_SIZE;
+                let end_frame = (entry.base + entry.length) as usize / FRAME_SIZE;
+                for frame in start_frame..end_frame {
+                    pfa.bitmap[frame / 8] &= !(1 << (frame % 8));
+                }
+            }
+        }
+        // set the bits corresponding to the bitmap as unavailable
+        let start_frame = core::ptr::addr_of!(pfa.bitmap) as usize / FRAME_SIZE;
+        let end_frame = (start_frame + pfa.bitmap.len()) / FRAME_SIZE;
+        for frame in start_frame..end_frame {
+            pfa.bitmap[frame / 8] |= 1 << (frame % 8);
+        }
+
+        pfa
+    }
+
+    pub fn allocate(&mut self) -> Result<PhysicalAddress, Error> {
+        for (byte_index, byte) in self.bitmap.iter_mut().enumerate() {
+            if *byte != 0xFF {
```

Comment:
Nit: Maybe add a comment to make this clear what this line is doing, something like `// if byte has a free frame`.

Reply:
I'll merge the existing code but improve the comments and add doc comments on the next commit.